docs: add contributing docs page (#6061)
software-mansion/react-native-reanimated: packages/docs-reanimated/docs/guides/contributing.mdx

---
id: contributing
title: Contributing
sidebar_label: Contributing
---

import useBaseUrl from '@docusaurus/useBaseUrl';

> Thank you for your interest in contributing to Reanimated! From commenting on and triaging issues, through extending the documentation, to reviewing and sending Pull Requests, all contributions are more than welcome.

Reviewer: ```suggestion
Thank you for your interest in contributing to Reanimated! From triaging and commenting on issues, through extending the documentation, to reviewing and sending Pull Requests, all contributions are more than welcome.
```

## Ways to Contribute

1. **Replying and handling open issues** &ndash; great way to contribute without writing a single line of code is triaging the issues. We often get issues that have generic errors, occur only in very specific cases, do not have proper example or reproducible repository. One of they way to help is preparing and filling those details, which will help other contributors get up to speed with the issue faster.

> 2. **Helping with documentation** &ndash; helping with the documentation can be as simple as checking for speling and grammer. If you encounter situations that can be explained better or in more details, click **Edit this page** at the bottom of most pages to get started with your own contribution. You can also help updating the page structures from old versions, as some of them are still ported from Reanimated v2 docs. Read more about [helping with documentation](#helping-with-documentation).

Reviewer: sure buddy

3. **Reviewing pull requests** &ndash; reviewing Pull Requests is crucial as it may help catch the corner cases or bugs that the contributor didn't notice. Every review matters as it may help us polish quality of the library.

> 4. **Contributing to Code** &ndash; contributions to the source code generally come in the form of Pull Requests. By contributing to the code you help us with solving issues, fixing bugs or introducing new, amazing features. If you want to start your adventure with open source it's good idea to take a look at [good first issue](https://github.com/software-mansion/react-native-reanimated/pulls?q=is%3Apr+is%3Aopen+label%3A%22good+first+issue%22) on GitHub. Read more about [contributing to code here](#contributing-to-code).

Reviewer 2: when you click _good first issue_, it takes us to PR section without any good first issue - is it on purpose? 😄

Author: Yeah, we don't have those yet, but I think idea is that they will be added :P

> 4. **Contributing to Code** &ndash; contributions to the source code generally come in the form of Pull Requests. By contributing to the code you help us with solving issues, fixing bugs or introducing new, amazing features. If you want to start your adventure with open source it's good idea to take a look at [good first issue](https://github.com/software-mansion/react-native-reanimated/pulls?q=is%3Apr+is%3Aopen+label%3A%22good+first+issue%22) on GitHub. Read more about [contributing to code here](#contributing-to-code).

Reviewer: ```suggestion
4. **Contributing to the Code** &ndash; contributions to the source code generally come in the form of Pull Requests. By contributing to the code you help us with solving issues, fixing bugs or introducing new, amazing features. If you want to start your adventure with open source it's good idea to take a look at [good first issue](https://github.com/software-mansion/react-native-reanimated/pulls?q=is%3Apr+is%3Aopen+label%3A%22good+first+issue%22) on GitHub. Read more about [contributing to code here](#contributing-to-code).
```

### Repository overview

The React Native Reanimated repository is a monorepo with the following structure:

```
├── apps
│   ├── common-app // shared source code of example apps
│   ├── paper-example // React Native app wrapper running the Old Architecture for shared examples
│   ├── fabric-example // React Native app wrapper running the New Architecture for shared examples
│   ├── macos-example // React Native for MacOS app wrapper for shared example code
│   ├── next-example // Next.js wrapper for shared example code
│   ├── tvos-example // React Native for TVOS app wrapper for shared examples
│   └── web-example // React Native for Web app wrapper for shared examples
└── packages

>     ├── docs-reanimated // documentation described further in [docs section](#helping-with-documentation)

Reviewer 2: when in code block the link is not linking - in a meaning it stays as a text and doesn't transform to link

>     ├── docs-reanimated // documentation described further in [docs section](#helping-with-documentation)

Author: Fixed in https://github.com/software-mansion/react-native-reanimated/pull/6061/commits/7a16a7185d4c979ad1aeacdb8f8398652f07d82d

    ├── eslint-plugin-reanimated // source of eslint plugin
    └── react-native-reanimated
        ├── android // source code of Android native implementation
        ├── apple // source code of iOS native implementation
        ├── Common // shared C++ code
        ├── scripts // JavaScript and bash scripts used in the CI pipeline
        ├── src // reanimated source code
        └── plugin // source code of the babel plugin
```

## Handling open issues

Understanding and reproducing the problem can be a very time consuming task. The great way to help other contributors get up to the speed with solving an issue is providing detailed description and _reproducible_ example. The GitHub already has an template for creating issue, nevertheless we still encounter issues that don't have all the necessary details, like:

- repository that we can clone and quickly see the problem,
- very generic reproduction steps,
- missing description,

> - not related or truncated stack trace.

Author: I just added `missing` to the first one, I think that tackle the problem. Fixed in 48019a86786763057939b208482bd11c2d50fd0f

> - not related or truncated stack trace.

Reviewer: Yes, that's good

1. Read [How to Open a Pull Request](https://reactnative.dev/contributing/how-to-open-a-pull-request)
2. Clone [react-native-reanimated](https://github.com/software-mansion/react-native-reanimated/)
3. `cd react-native-reanimated`
4. `yarn`
5. `cd packages/docs-reanimated`
6. `yarn && yarn start`

This will start your local Docusaurus server running the documentation.

### Project structure

Here you can find the most important parts of the `react-native-reanimated/packages/docs-reanimated` project. Not all folders are listed for brevity.

```
├── docs // contains folder with md/mdx files, each folder is a section in the side navigation, and each file is documentation page
├── src
│   ├── components // contains components used to make the pages interactive
│   └── examples // contains examples that are later presented in the documentation
└── static
    ├── img // contains images used in the docs
    └── recordings // contains video recordings of the examples

> ```

Reviewer: Looks a bit strange when scrolled with these buttons over it - they are kinda messing with each other. (cc @patrycjakalinska @kacperkapusciak )
 
https://github.com/software-mansion/react-native-reanimated/assets/77503811/0f4deebe-8c1e-4c8e-89c2-a826f7308030

Reviewer: You could also make the sentence shorter - I think it would do the job

Reviewer 3: i prefer to have it as a tree structure 

yes, you can add a prop but I don't think these buttons are a real issue tho

Author: Ok, so I'm leaving as it is

### Documentation page structure

> It's important that the structure of the site be consistent, as this helps developers to navigate easily and quickly find what they are looking for. In v3 documentation, we assume a given page structure:

Reviewer: ```suggestion
It's important that the structure of the site be consistent, as this helps developers navigate easily and quickly find what they are looking for. In v3 documentation, we assume a given page structure:
```

### Documentation writing style guide

The developers want to find the solution to their problems as fast as possible. The speed of finding answers is the reason we optimize the documentation for [skimming](https://www.linkedin.com/pulse/skimming-vs-speed-reading-whats-difference-paul-nowak#:~:text=Skimming%20is%20a%20method%20of,have%20to%20read%20every%20word.). We believe that improvements to documentation have direct impact on the overall developer experience. This section contains a couple of good practices when writing a technical documentation. Treat these rules more like guidelines than the source of absolute truth.

If you have spare 2 hours, we highly recommend taking a free [Technical Writing One](https://developers.google.com/tech-writing/one/) course by Google.

##### Avoid passive voice

Most of the sentences in a technical documentation should be in active voice. Sentences in the active voice are usually simpler than the ones in the passive voice.

See the following examples:

- The arguments **are provided to** the function.

> - A worklet **are generated by** a babel plugin.

Reviewer: ```suggestion
- A worklet **is generated by** the babel plugin.
```

- UI thread **is being used by** Reanimated.

which if we rewrite these sentences to active voice become simpler:

> - A babel plugin **generates** worklet.
> - The function **receives** the arguments.
> - Reanimated **runs** on the UI thread.

Reviewer 4: ```suggestion
- The function **receives** the arguments.
- A babel plugin **generates** worklet.
- Reanimated **runs** on the UI thread.
```

> - A babel plugin **generates** worklet.
> - The function **receives** the arguments.
> - Reanimated **runs** on the UI thread.

Reviewer: You shuffled the order 😄 
```suggestion
- The function **receives** the arguments.
- The babel plugin **generates** a worklet.
- Reanimated **runs** on the UI thread.
```

Acronyms can often lead to confusion.

For example, in the context of libraries, abbreviations like `REA`, `RNGH`, `RNS`, may be obvious to developers proficient in the React Native open source space, but completely unknown to beginners.

Some acronyms, such as RN, can mean both React Native and React Navigation.

When explaining topics that often use domain-specific acronyms, such as `JSI`, `JNI`, `JSC`, you should first introduce them. For example, JavaScript Core (JSC) or JavaScript Interface (JSI).

The following rule doesn't apply for a globally established abbreviation like `UI`, `API`, `HTML`, `TS` etc.

#### Use contractions

Prefer writing using contractions like `don't`, `doesn't`, `it's` instead of full forms like `do not`, `does not`, `it is`.

### Embedding interactive examples

We try to make the code examples as easy as it gets. This is why good, interactive examples are crucial. To create interactive examples in the documentation you can use `InteractiveExample` component. It takes `src` prop that expects source of a component that you can declare in `src/examples`, and a `component` prop that showcases given code.

#### Example with a video

```javascript

> import useBaseUrl from '@docusaurus/useBaseUrl';

Reviewer: is this import important in this example?

Author: Fixed in 0b194ed38b115747e70e2cf02f6c5979ac0ae730

import AnimatedKeyboardSrc from '!!raw-loader!@site/src/examples/AnimatedKeyboard';

<InteractiveExample
  src={AnimatedKeyboardSrc}
  component={
    <ThemedVideo
      center
      width={300}
      sources={{
        light: '/recordings/useAnimatedKeyboard_light.mov',
        dark: '/recordings/useAnimatedKeyboard_dark.mov',
      }}
    />
  }
/>;
```

#### Example with an interactive component

```javascript
import DecayTrain from '@site/src/examples/DecayTrain';
import DecayTrainSrc from '!!raw-loader!@site/src/examples/DecayTrain';

<InteractiveExample
  src={DecayTrainSrc}
  component={<DecayTrain />}
  label="Grab and drag the train"
/>;
```

> ## Contributing to Code

Reviewer: ```suggestion
## Contributing to the Code
```

> Posting Pull Requests to the issues is great way to contribute to Reanimated. If you eager to start contributing right away, we have list of [good first issue](https://github.com/software-mansion/react-native-reanimated/pulls?q=is%3Apr+is%3Aopen+label%3A%22good+first+issue%22) that contain bugs which have limited scope. In this section we'll describe in more details how to play around with Reanimated setup.

Reviewer: ```suggestion
Posting Pull Requests to the issues is a great way to contribute to Reanimated. If you are eager to start contributing right away, we have a list of [good first issues](https://github.com/software-mansion/react-native-reanimated/pulls?q=is%3Apr+is%3Aopen+label%3A%22good+first+issue%22) that contain bugs which have limited scope. In this section we'll describe how to play around with Reanimated setup in more detail.
```

:::caution

> Make sure that you have [React Native setup](https://reactnative.dev/docs/environment-setup) properly and you got familiar with [How to Open a Pull Request](https://reactnative.dev/contributing/how-to-open-a-pull-request) before reading this section further.

Reviewer: ```suggestion
Make sure that you have [React Native setup](https://reactnative.dev/docs/environment-setup) properly and you are familiar with [How to Open a Pull Request](https://reactnative.dev/contributing/how-to-open-a-pull-request) before reading this section further.
```

:::

We have shared code of example app, that is run on different platforms and architectures. When working on an issue the most important destinations will be:

- `apps/common-app` &ndash; shared source code with example app
- `packages/react-native-reanimated/src/reanimated2` &ndash; containing source code related to JS part of Reanimated features
- `packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated` &ndash; containing source code related to Android native part of Reanimated features
- `packages/react-native-reanimated/apple` &ndash; containing source code related to iOS native part of Reanimated features

> Our workflow usually starts from editing `app/src/examples/EmptyExample.tsx` with testing example for our new feature or a bug. This is the part that we often attach to PR so other developers can test the same use-case. Next, we run the `Example` (or `FabricExample` depending on the architecture) app and start development.

Reviewer: It seems a bit unclear and suggests commiting the code to the empty example 😆 
```suggestion
Our workflow usually starts from editing `app/src/examples/EmptyExample.tsx` with testing example for our new feature or a bug. This testing code from `EmptyExample` is then often attached to the PR so other developers can test the same use-case (remember not to commit this code in the `EmptyExample` file!). Next, we run the `Example` (or `FabricExample` depending on the architecture) app and start development.
```

:::info

> If you want implement new feature or fix bug, but still feel uninformed, after reading through this section, or could use a guidance you can reach Reanimated Team on [discord](https://discord.com/channels/464786597288738816/1216675825584181330).

Reviewer: `Uninformed` is a pretty difficult word I'd say 
```suggestion
If you want to implement a new feature or fix a bug, but still are unsure after reading through this section or could use some guidance, you can reach Reanimated Team on [discord](https://discord.com/channels/464786597288738816/1216675825584181330).
```

:::

> ### Working on Android

Reviewer: ```suggestion
### Working with Android
```

To start with, let's install all dependencies:

1. `yarn && yarn build`
2. `cd apps/common-app && yarn && cd ..`

> 3. `cd paper-example && yarn`

Reviewer: As far as I know we don't need to use `yarn` in neither `common-app` nor `paper-example` anymore because we have hoisted deps from monorepo structure (but please confirm it @tjzel )

Author: fixed in https://github.com/software-mansion/react-native-reanimated/pull/6061/commits/7a16a7185d4c979ad1aeacdb8f8398652f07d82d

4. `yarn start` &ndash; make sure to start metro bundler before building the app in Android Studio.

and open `react-native-reanimated/apps/paper-example/android` with Android Studio.

<img src={useBaseUrl('/img/contributing/android_studio.png')} />

> The native source code of Reanimated can be found in `react-native-reanimated` module, in `/packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated`. Making sure the Metro builder is running. You can now build React Native app or debug native code.

Reviewer: ```suggestion
The native source code of Reanimated can be found in `react-native-reanimated` module, in `/packages/react-native-reanimated/android/src/main/java/com/swmansion/reanimated`. Make sure the Metro bundler is running. You can now build React Native app or debug native code.
```

Project structure overview:

```
├── AndroidUIScheduler.java
├── CopiedEvent.java
├── MapUtils.java
├── NativeMethodsHelper.java
├── NodesManager.java
├── ReactNativeUtils.java
├── ReanimatedMessageQueueThreadBase.java
├── ReanimatedPackage.java
├── ReanimatedUIManagerFactory.java
├── Utils.java
├── keyboard
├── layoutReanimation
├── nativeProxy
└── sensor

> ```

Reviewer 4: I am unsure if it is necessary; potentially, those names could change in the future.

Author: And at the same time it's not very informative unless we want to describe this files more. I'd go with removing those now and consider upgrading the focus with more details if we decide to @kacperkapusciak what do you think?

Reviewer 3: i'm okay with removing the structure. pointing people to the correct path where source is is good enough

> ### Working on iOS

Reviewer: ```suggestion
### Working with iOS
```

To begin with, let's install all the necessary dependencies:

> 1. `yarn && yarn build`
> 2. `cd apps/common-app && yarn && cd ..`
> 3. `cd paper-example && yarn`
> 4. `yarn start` &ndash; make sure to start metro bundler before building the app in Xcode.

Reviewer: It's kinda wrong. Firstly, we don't need yarn in `common-app` or `paper-example`. Secondly, we need to install the pods.
```suggestion
1. `yarn && yarn build`
2. `cd paper-example/ios && bundle install && bundle exec pod install` &ndash; install Pods for iOS
3. `cd .. && yarn start` &ndash; make sure to start metro bundler before building the app in Xcode.
```

and open `react-native-reanimated/apps/paper-example/ios/ReanimatedExample.xcworkspace` with Xcode.

<img src={useBaseUrl('/img/contributing/xcode.png')} />

> To find the native source code of Reanimated navigate to `Pods > Development Pods >  RNReanimated > Example > node_modules > react-native-reanimated > apple`. Making sure the Metro builder is running. You can now build React Native app or debug native code.

Reviewer: ```suggestion
To find the native source code of Reanimated navigate to `Pods > Development Pods >  RNReanimated > Example > node_modules > react-native-reanimated > apple`. Make sure the Metro bundler is running. You can now build React Native app or debug native code.
```

Project structure overview:

```sh
├── Fabric
├── LayoutReanimation
├── RCTEventDispatcher+Reanimated.h
├── RCTEventDispatcher+Reanimated.m
├── RCTUIView+Reanimated.h
├── RCTUIView+Reanimated.m
├── READisplayLink.h
├── REAModule.h
├── REAModule.mm
├── REANodesManager.h
├── REANodesManager.mm
├── REASlowAnimations.h
├── REASlowAnimations.mm
├── REAUIKit.h
├── REAUtils.h
├── REAUtils.m
├── RNGestureHandlerStateManager.h
├── RNReanimated.xcodeproj
├── keyboardObserver
├── native
└── sensor
```

> ### Preparing Pull Request

Reviewer: ```suggestion
### Preparing a Pull Request
```

When the code changes are ready, it's time to open a Pull Request. We configured the repository to come with a template to help you post your changes properly. The most important of these are:

1. **Summary**:

> - If you're solving specific issue, start by linking it, for example: Fixes #2137.

Reviewer: ```suggestion
- If you're solving a specific issue, start by linking it, for example: Fixes #2137 (here goes a full link to an issue and GitHub shortens it into a #number format).
```

- Write what are your motivations and how you solved the problem.

> - If applicable, it's a good idea to include a screen or video before and after making changes.

Reviewer: ```suggestion
- If applicable, it's a good idea to include a screenshot or a video of before and after making changes.
```

> 2. **Test plan** &ndash; Describe how others can test your change, here we usually attach the code from your modified `EmptyExample`, that you used for testing.

Reviewer: ```suggestion
2. **Test plan** &ndash; Describe how others can test your changes, here we usually attach the code from your modified `EmptyExample`, that you used for testing.
```
